### PR TITLE
Add extension methods for creating OptionsBuilder with ValidateOnStart support

### LINKED ISSUE #89263: [API Proposal]: Add extensions to create an OptionsBuilder with ValidateOnStart support
### Background and motivation

Based on cloud native scenarios, these extensions methods were used and they make sense to add to the System.Microsoft.Extensions.DependencyInjection library.

Note that in cloud native, these methods were called "AddValidatedOptions"; here they are called "AddValidateOnStartOptions".

### API Proposal

```diff
namespace namespace Microsoft.Extensions.DependencyInjection;

public static partial class OptionsBuilderExtensions
{
    // The one existing extension method:
    public static OptionsBuilder<TOptions> ValidateOnStart
        <[DynamicallyAccessedMembers(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.
            PublicParameterlessConstructor)] TOptions>
        (this OptionsBuilder<TOptions> optionsBuilder)
        where TOptions : class;

    // New one to call ValidateOnStart() above
+   public static OptionsBuilder<TOptions> AddValidateOnStartOptions
+       <[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] TOptions>
+        (this IServiceCollection services,
+        string? name = null)
+        where TOptions : class
    {
        // Simple wrapper like this:
        return new OptionsBuilder<TOptions>(services, name ?? Options.DefaultName).ValidateOnStart();
    }

    // New one to support IValidateOptions and call ValidateOnStart() above
+   public static OptionsBuilder<TOptions> AddValidateOnStartOptions<
+       [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] TOptions,
+       [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TValidateOptions>(
+       this IServiceCollection services,
+       string? name = null)
+       where TOptions : class
+       where TValidateOptions : class, IValidateOptions<TOptions>
      {
         // Simple wrapper like this:
         services.AddOptions().TryAddEnumerable(ServiceDescriptor.Singleton<IValidateOptions<TOptions>, TValidateOptions>());
         return new OptionsBuilder<TOptions>(services, name ?? Options.DefaultName).ValidateOnStart();
      }
```

### API Usage

```csharp
// Before:
IHostBuilder builder = ...
builder.services
    .AddOptions<MyOptions>()
    .Bind(builder.Configuration.GetSection(MyConfigOptions.MyConfig))
    .Validate(o => o.MySetting == true)
    .ValidateOnStart();
// After:
IHostBuilder builder = ...
builder.Services
    .AddValidateOnStartOptions<MyOptions>()
    .Bind(builder.Configuration.GetSection(MyConfigOptions.MyConfig))
    .Validate(o => o.MySetting == true);

// Before:
IHostBuilder builder = ...
builder.Services
    .AddOptions<MyOptions>()
    .AddSingleton<IValidateOptions<MyOptions>>(new MyOptionsValidator(builder .Name)) // or TryAddEnumerable
    .Bind(builder.Configuration.GetSection(MyConfigOptions.MyConfig))
    .ValidateOnStart();
// After:
IHostBuilder builder = ...
builder.Services.
    .AddValidateOnStartOptions<MyOptions, MyOptionsValidator>(builder.Name))
    .Bind(builder.Configuration.GetSection(MyConfigOptions.MyConfig))
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/89263

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @halter73

**Review by tarekgh — APPROVED** [review]
LGTM.

**Comment by tarekgh** [comment]
CC @ericstj

File: src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -31,6 +31,43 @@ public static IServiceCollection AddOptions(this IServiceCollection services)
...
+            this IServiceCollection services,
+            string? name = null)
+            where TOptions : class
+        {
+            return new OptionsBuilder<TOptions>(services, name ?? Options.Options.DefaultName).ValidateOnStart();
+        }
+
+        /// <summary>
+        /// Adds services required for using options and enforces options validation check on start rather than in runtime.
+        /// The <seealso cref="OptionsBuilderExtensions.ValidateOnStart{TOptions}(OptionsBuilder{TOptions})"/> extension is called by this method.
```
Comment (geeknoid) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:53]: Same

File: src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -31,6 +31,43 @@ public static IServiceCollection AddOptions(this IServiceCollection services)
...
+        /// </summary>
+        /// <typeparam name="TOptions">The options type to be configured.</typeparam>
+        /// <typeparam name="TValidateOptions">The <see cref="IValidateOptions{TOptions}"/> validator type.</typeparam>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+        /// <param name="name">The name of the options instance.</param>
+        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+        public static OptionsBuilder<TOptions> AddOptionsWithValidateOnStart<
```
Comment (geeknoid) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:60]: This doesn't exactly roll off the tongue.

Would it be too late to name it something cleaner? Say "AddEagerlyValidatedOptions" perhaps?

**Review by geeknoid — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs (hunk 2 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:60]: During API review, the thought was starting with "AddOptions" was preferrable due to intellisense suggestions.

File: src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -31,6 +31,43 @@ public static IServiceCollection AddOptions(this IServiceCollection services)
...
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] TOptions>(
+            this IServiceCollection services,
+            string? name = null)
+            where TOptions : class
+        {
+            return new OptionsBuilder<TOptions>(services, name ?? Options.Options.DefaultName).ValidateOnStart();
```
Comment (halter73) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:48]: Shouldn't this call `AddOptions`? I think neither the `OptionsBuilder` constructor nor `ValidateOnStart` do this.

```suggestion
            services.AddOptions();
            return new OptionsBuilder<TOptions>(services, name ?? Options.Options.DefaultName).ValidateOnStart();
```

**Review by halter73 — COMMENTED** [review]
We technically approved adding this to `OptionsBuilderExtensions` instead of `OptionsServiceCollectionExtensions` in API review for some reason, but `OptionsServiceCollectionExtensions` is the correct place.

File: src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -31,6 +31,43 @@ public static IServiceCollection AddOptions(this IServiceCollection services)
             return services;
         }
 
+        /// <summary>
+        /// Adds services required for using options and enforces options validation check on start rather than in runtime.
+        /// The <seealso cref="OptionsBuilderExtensions.ValidateOnStart{TOptions}(OptionsBuilder{TOptions})"/> extension is called by this method.
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:36]: ping @geeknoid - please clarify comment above. Thanks

File: src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:50]: `ValidateOnStart()` does call `optionsBuilder.Services.AddOptions<StartupValidatorOptions>()`

File: src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -31,6 +31,47 @@ public static IServiceCollection AddOptions(this IServiceCollection services)
...
+        /// <remarks>
+        /// The <seealso cref="OptionsBuilderExtensions.ValidateOnStart{TOptions}(OptionsBuilder{TOptions})"/> extension is called by this method.
+        /// </remarks>
+        /// <typeparam name="TOptions">The options type to be configured.</typeparam>
+        /// <typeparam name="TValidateOptions">The <see cref="IValidateOptions{TOptions}"/> validator type.</typeparam>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+        /// <param name="name">The name of the options instance.</param>
+        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+        public static OptionsBuilder<TOptions> AddOptionsWithValidateOnStart<
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] TOptions,
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:65]: Why does this need to be `.All`? Shouldn't it be

```C#
[DynamicallyAccessedMembers(Options.DynamicallyAccessedMembers)] TOptions,
```

?

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:65]: The same question for the other overload as well. I believe it should be:

```C#
public static OptionsBuilder<TOptions> AddOptionsWithValidateOnStart<
            [DynamicallyAccessedMembers(Options.DynamicallyAccessedMembers)] TOptions>(
            this IServiceCollection services,
            string? name = null)
```

**Pull request merged**